README.md
# ToyLane
Toy Lane Detection Repo



history/history.txt
{"meta": {"test": "data/datasets/single_lane_front_test.lmdb", "notes": "6 anchors. 1500 train with 10 distortions.  front view. adam.", "train": "data/datasets/single_lane_front_train.lmdb", "num_train": null, "num_test": null}, "params": {"metrics": ["mean_absolute_error"], "loss": "mean_squared_error", "num_train": null, "optimizer": "adam", "fc_init": "uniform", "num_test": null, "nb_epoch": 10, "batch_size": 32}, "name": "model_0"}
{"meta": {"test": "data/datasets/single_lane_front_test.lmdb", "notes": "6 anchors. 1500 train with 10 distortions.  front view. adam.", "train": "data/datasets/single_lane_front_train.lmdb", "num_train": null, "num_test": null}, "params": {"metrics": ["mean_absolute_error"], "loss": "mean_squared_error", "num_train": null, "optimizer": "adam", "fc_init": "uniform", "num_test": null, "nb_epoch": 10, "batch_size": 32}, "name": "model_1"}
{"meta": {"test": "data/datasets/single_lane_topdown_train.lmdb", "notes": "2 lane 6 anchors, top down, learning rate decreases by .5 when learning halts. regressing displacements from center.", "train": "data/datasets/single_lane_topdown_train.lmdb", "num_train": null, "num_test": null}, "params": {"model_arch": "model_2_lane_6_anchor_regression", "loss": "mean_squared_error", "num_train": null, "optimizer": "adam", "fc_init": "uniform", "metrics": ["mean_absolute_error"], "num_test": null, "nb_epoch": 7, "batch_size": 32}, "name": "model_2"}
{"meta": {"test": "data/datasets/single_lane_topdown/test.lmdb", "notes": "same as model2 but with 20 epochs to get learning rate to change", "train": "data/datasets/single_lane_topdown/train.lmdb", "num_train": null, "num_test": null}, "params": {"model_arch": "model_2_lane_6_anchor_regression", "loss": "mean_squared_error", "num_train": null, "optimizer": "adam", "fc_init": "uniform", "metrics": ["mean_absolute_error"], "num_test": null, "nb_epoch": 20, "batch_size": 32}, "name": "model_3"}
{"meta": {"test": "data/datasets/single_lane_topdown_2_anchor/test.lmdb", "notes": "2 anchor 2 lane", "train": "data/datasets/single_lane_topdown_2_anchor/train.lmdb", "num_train": null, "num_test": null}, "params": {"model_arch": "model_2_lane_2_anchor_regression", "loss": "mae", "num_train": null, "optimizer": "adam", "fc_init": "he_normal", "metrics": ["mae"], "num_test": null, "nb_epoch": 30, "batch_size": 32}, "name": "model_4"}


simplelmdb.py
import lmdb
import msgpack
import msgpack_numpy as m
m.patch()
import numpy as np
import random
import shutil
import os
class Simplelmdb(object):
    def __init__(self, path):
        self.path = path
        self.db = lmdb.open(path, map_size=10**13)

    def items(self):
        with self.db.begin() as txn:
            cur = txn.cursor()
            cur.set_range('')
            for k, v in cur:
                yield k, msgpack.unpackb(v)

    def keys(self):
        return (k for k,v in self.items())

    def values(self):
        return (v for k,v in self.items())

    def get(self, k, *args, **kwargs):
        with self.db.begin() as txn:
            val = txn.get(k)
            if val is None and len(args) > 0:
                return args[0]
            return msgpack.unpackb(val)

    def get_keys(self, keys, *args, **kwargs):
    	"""
    	second arg can be the default
    	"""
	with self.db.begin() as txn:
		for k in keys:
			val = txn.get(k)
			if val is None and len(args) > 0:
				yield args[0]
			yield msgpack.unpackb(val)


    def put(self, k, v):
        with self.db.begin(write=True) as txn:
            val = msgpack.packb(v)
            txn.put(k, val)

    def count(self):
        return self.db.stat()['entries']

    def take(self, n):
        items = self.items()
        for i in range(n):
            yield items.next()

    def first(self):
        return self.items().next()

model_history.py
import argparse
import pickle
import time
import json
import os
import shutil
import glob
class ModelHistory(object):
	def __init__(self, history_dir, name, meta, params):
		# directory containing all our models
		self.history_dir = history_dir
		self.notes_file = history_dir + '/history.txt'

		self.meta = meta
		self.params = params

		self.create_history_dir()
		self.model_number = self.get_model_number()
		self.model_name = name
		self.model_dir = self.history_dir + '/' + self.model_name + "_" + self.model_number
		self.model_save = self.model_dir + '/model'
		self.notes_save = self.model_dir + '/notes.json'
		self.training_history = self.model_dir + '/training_history.pkl'
		self.predictions = self.model_dir + '/predictions.pkl'
		self.tensorboard = self.model_dir + '/tensorboard'
		self.update_model_history()
		self.create_model_dir()


	def create_history_dir(self):
		if not os.path.exists(self.history_dir):
			os.makedirs(self.history_dir)

	def get_model_number(self):

		if not os.path.exists(self.notes_file):
			return 0        	
		old_model_numbers = []
		with open(self.notes_file) as f:
			for line in f:
				line = line.strip()
				j = json.loads(line)
				if j['name'] == self.name:
					old_model_numbers.append(j['model_number'])
				
		if not old_model_numbers:
			return 0

		return max(old_model_numbers) + 1

	def update_model_history(self):
		history = self.get_history()
		with open(self.notes_file, 'a') as f:
			line = json.dumps(history)
			f.write(line + '\n')


	def get_history(self):
		return {
			'name' : self.model_name,
			'number' : self.number,
			'meta' : self.meta,
			'params' : self.params
		}

	def create_model_dir(self):
		if os.path.exists(self.model_dir):
			raise Exception("Something went wrong, this model already exists: " + self.model_dir)
		os.makedirs(self.model_dir)  
		history = self.get_history() 
		with open(self.notes_save, 'w') as f:
			line = json.dumps(history)
			f.write(line + '\n')

create_dataset.py
import numpy as np
import pickle
import cv2
from simplelmdb import Simplelmdb
import lmdb
import numpy as np
import lmdb
import msgpack
import msgpack_numpy as m
m.patch()
import random
import shutil
import os

def get_inverse_perspective_transform(): 
    return Controller.get_perspective_transform(inverse=True)


def get_perspective_transform(inverse=False):        
    p1 = (232, 380) # top left
    p2 = (360, 380) # top right
    p3 = (206, 414) # bottom left
    p4 = (383, 414) # bottom right

    pts1 = np.float32([list(p1),list(p2),list(p3),list(p4)])
    output_width = 100
    output_height = 35
    dx=289
    dy=585
    pts2 = np.float32([[dx,dy],[dx+output_width,dy],[dx,dy+output_height],[dx+output_width,dy+output_height]])
    if inverse:
        M = cv2.getPerspectiveTransform(pts2,pts1)
    else:
        M = cv2.getPerspectiveTransform(pts1,pts2)
    return M


def transform_anchor_points(points, M):
    points = np.array(points, dtype=np.float32)
    transformed = cv2.perspectiveTransform(points[None, :, :], M)
    return transformed.tolist()[0]


def compute_topdown_image(image_filename, transformation_matrix, transformation_size):
    img = cv2.imread(image_filename)
    img = cv2.cvtColor(img, cv2.cv.CV_BGR2RGB)
    dst = cv2.warpPerspective(img, transformation_matrix, (transformation_size , transformation_size))
    dst = cv2.cvtColor(dst, cv2.cv.CV_BGR2RGB)
    return dst


def crop(image, anchor_points, crop_size, shift_percentage):
    """
    shift is shift from center
    """
    image_width = image.shape[1]
    max_shift = (image_width - crop_size[1])/2.0
    
#     print "maxshift"
#     print max_shift
    # left x for when the crop window is centered
    x_left_image_center = max_shift
    shift = shift_percentage*max_shift
    
    # max
    x_left = x_left_image_center + shift
    x_right = x_left+crop_size[1]
    

    anchor_points = np.copy(anchor_points)
    anchor_points[:,0] = anchor_points[:, 0] - (x_left)
    
#     print x_left
#     print shift
    cropped = np.copy(image[:crop_size[0],x_left:x_right])
    return cropped, anchor_points

    #crop the center of the image    
def resize(image, anchor_points, scale):
    anchor_points = anchor_points * scale
    image = cv2.resize(image, None, fx=scale, fy=scale, interpolation = cv2.INTER_CUBIC)
    return image, anchor_points

def change_perspective(image, anchor_points, transformation, transformation_size):
    anchor_points = cv2.perspectiveTransform(anchor_points[None, :, :], transformation)[0]
    image = cv2.warpPerspective(image, transformation, transformation_size)
    return image, anchor_points
    
def get_anchor_points(source, image_file, use_front=False):
    splines = [lane['anchor_points'] for lane in source['lanes'][image_file].values()]
    anchor_points = [anchor_point for spline in splines for anchor_point in spline ]
    anchor_points = 1.0*np.array(anchor_points)
#     print '1'
#     print anchor_points    
    if use_front:
        transformation = topdown_to_front_transformation = source['topdown_to_front_transformation']
        anchor_points = cv2.perspectiveTransform(anchor_points[None, :, :], transformation)[0]
#     print '2'
#     print anchor_points          
    return anchor_points
        
def calculate_6_anchor_displacements(image, anchor_points):
    center_x = image.shape[1]/2.0
    displacements = anchor_points[:,0] - center_x
    return displacements

def calculate_2_anchor_displacements(image, anchor_points):
    anchor_points = anchor_points[[2,5],:]
    center_x = image.shape[1]/2.0
    displacements = anchor_points[:,0] - center_x
    return displacements

def calculate_2_anchor_pixel_classification(image, anchor_points):
    anchor_points = anchor_points[[2,5],:]
    classes = p.zeros(image.shape[1])
    for i in anchor_points.shape[0]:
        anchor_point = get(get_pixel_value, anchor_points[i,:])
        classes[anchor_point[0]] = 1.0
    return classes

def get_pixel_value(image, anchor_point):
    x = min(max(0,int(anchor_point[0])), image.shape[1])
    y = min(max(0,int(anchor_point[1])), image.shape[0])
    return np.array([x,y])

def create_image_data(image, anchor_points, size, shift_percentage, targets_fn):
    scale = (1.0*size[0])/image.shape[0]
    image, anchor_points = resize(image, anchor_points, scale)
#     print 'C'
#     print anchor_points    
    image, anchor_points = crop(image, anchor_points, size, shift_percentage)
#     print 'D'
#     print anchor_points     
    targets = targets_fn(image, anchor_points)
    image = 1.0*np.expand_dims(image, axis=0)
    return image, anchor_points, targets

TARGETS_FN = {
    'calculate_2_anchor_pixel_classification' : calculate_2_anchor_pixel_classification,
    'calculate_6_anchor_displacements' : calculate_6_anchor_displacements,
    'calculate_2_anchor_displacements' : calculate_2_anchor_displacements,
    'calculate_2_anchor_pixel_classification' : calculate_2_anchor_pixel_classification
}

def create_dataset(save_file, images, source, targets_fn, num_features, size=(224,224), shuffle=False, data_max=None, data_min=None, num_shifts=1, use_front=False, normalize_targets=False):    
    n = len(images)*(1+num_shifts)
    map_size = n * size[0]*size[1]*8 * 20

    if os.path.exists(save_file):
        shutil.rmtree(save_file)

    env = lmdb.open(save_file, map_size=map_size)
    
    transformation_size = source['transformation_size']
    front_to_topdown_transformation = source['front_to_topdown_transformation']
    topdown_to_front_transformation = source['topdown_to_front_transformation']  
    
    targets = np.zeros((n, num_features))
    with env.begin(write=True) as txn:
        for i in range(len(images)):
            shift_percentages = (np.random.random(num_shifts+1)*2)-1
            shift_percentages[0] = 0
            image_file = images[i]
            image = cv2.imread(image_file)
            image = cv2.cvtColor(image, cv2.cv.CV_BGR2RGB)
            anchor_points = get_anchor_points(source, image_file, use_front=use_front)
            if not use_front:
                image, _ = change_perspective(image, anchor_points, front_to_topdown_transformation, transformation_size)
            for j in range(len(shift_percentages)):

                sample_anchor_points = np.copy(anchor_points)
                shift_percentage = shift_percentages[j]
                image_data, sample_anchor_points, displacements = create_image_data(image, sample_anchor_points, size, shift_percentage, targets_fn)  
                image_index = i*(1+num_shifts) + j
                targets[image_index,:] = displacements
                sample = {
                    'image': image_file,
                    'image_data' : image_data[0], 
                    'image_targets' : displacements, 
                    'shift_percentage' : shift_percentage,
                    'anchor_points' : sample_anchor_points
                }
                sample_encoded = msgpack.packb(sample, default=m.encode)
                txn.put(str(image_index), sample_encoded)
                if image_index % 1000 ==0:
                    print image_index
                    env.sync()

        low = 0.1
        high = 0.9    
        data_min = None
        data_max = None

        if normalize_targets:                
            print "Normalizing"
            normalized_targets, data_min, data_max = normalize_to_range(targets, low, high, data_min=data_min, data_max=data_max)

            for image_index in range(n):
                key = str(image_index)
                sample_encoded = txn.get(key)            
                sample = msgpack.unpackb(sample_encoded, object_hook=m.decode)
                sample['image_targets'] = normalized_targets[image_index]
                sample_encoded = msgpack.packb(sample, default=m.encode) 
                txn.put(key, sample_encoded)

    db = Simplelmdb(save_file)
    db.put('num_records', n)
    if normalize_targets:       
        db.put('low', low)
        db.put('high', high)
        db.put('data_max', data_max)
        db.put('data_min', data_min)

    return data_min, data_max
            
def normalize_to_range(data,low, high, data_min=None, data_max=None):
    # Normalize to [0, 1]:
    if data_min is None:
        data_min = np.min(data, axis=0)
    if data_max is None:
        data_max = np.max(data, axis=0)
    data_range = data_max-data_min
    data = (data - data_min) / data_range;
    # Then scale to [x,y]:
    target_range = high - low;
    data = (data*target_range) + low;
    return data, data_min, data_max

def unnormalize(data, low, high, data_min, data_max):
    target_range = high - low;
    data = (data -low)/target_range
    data_range = data_max-data_min
    data = (data*data_range) + data_min
    return data
    


finetune_resnet_dd.py
import argparse
from resnet50 import ResNet50
from keras.preprocessing import image
from imagenet_utils import preprocess_input, decode_predictions
from keras.models import Sequential
from keras.layers import Activation, Dropout, Flatten, Dense, Input
from keras.callbacks import ModelCheckpoint, Callback, EarlyStopping, TensorBoard
import pickle
from keras.models import Model
from keras import backend as K
import keras
import numpy as np
import time
from model_history import ModelHistory
import random
from simplelmdb import Simplelmdb
import itertools as it
import numpy as np
import json


if __name__ == '__main__':
	parser = argparse.ArgumentParser()
	parser.add_argument("--data", required=True,help="lmdb folder containing train data")
	parser.add_argument("--model_arch", required=True, help="architecture of the model to use")
	parser.add_argument("--loss", required=True, help="architecture of the model to use")
	parser.add_argument("--metric", required=True, help="architecture of the model to use")
	parser.add_argument("--optimizer", required=True, help="architecture of the model to use")
	parser.add_argument("--fc_init", required=True, help="architecture of the model to use")
	parser.add_argument("--num_train", type=int, help="recoring location")
	parser.add_argument("--num_test", type=int, help="recoring location")
	parser.add_argument("--num_epochs",default=7, type=int, help="recoring location")
	parser.add_argument("--history", help="folder containing model history.")
	args = parser.parse_args()

class ReduceLROnPlateau(Callback):
    '''Reduce learning rate when a metric has stopped improving.
    Models often benefit from reducing the learning rate by a factor
    of 2-10 once learning stagnates. This callback monitors a
    quantity and if no improvement is seen for a 'patience' number
    of epochs, the learning rate is reduced.
    # Example
        ```python
            reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2,
                                          patience=5, min_lr=0.001)
            model.fit(X_train, Y_train, callbacks=[reduce_lr])
        ```
    # Arguments
        monitor: quantity to be monitored.
        factor: factor by which the learning rate will
            be reduced. new_lr = lr * factor
        patience: number of epochs with no improvement
            after which learning rate will be reduced.
        verbose: int. 0: quiet, 1: update messages.
        mode: one of {auto, min, max}. In `min` mode,
            lr will be reduced when the quantity
            monitored has stopped decreasing; in `max`
            mode it will be reduced when the quantity
            monitored has stopped increasing; in `auto`
            mode, the direction is automatically inferred
            from the name of the monitored quantity.
        epsilon: threshold for measuring the new optimum,
            to only focus on significant changes.
        cooldown: number of epochs to wait before resuming
            normal operation after lr has been reduced.
        min_lr: lower bound on the learning rate.
    '''

    def __init__(self, monitor='val_loss', factor=0.1, patience=10,
                 verbose=0, mode='auto', epsilon=1e-4, cooldown=0, min_lr=0):
        super(Callback, self).__init__()

        self.monitor = monitor
        if factor >= 1.0:
            raise ValueError('ReduceLROnPlateau does not support a factor >= 1.0.')
        self.factor = factor
        self.min_lr = min_lr
        self.epsilon = epsilon
        self.patience = patience
        self.verbose = verbose
        self.cooldown = cooldown
        self.cooldown_counter = 0  # Cooldown counter.
        self.wait = 0
        self.best = 0
        self.mode = mode
        self.monitor_op = None
        self.reset()

    def reset(self):
        if self.mode not in ['auto', 'min', 'max']:
            warnings.warn('Learning Rate Plateau Reducing mode %s is unknown, '
                          'fallback to auto mode.' % (self.mode), RuntimeWarning)
            self.mode = 'auto'
        if self.mode == 'min' or (self.mode == 'auto' and 'acc' not in self.monitor):
            self.monitor_op = lambda a, b: np.less(a, b - self.epsilon)
            self.best = np.Inf
        else:
            self.monitor_op = lambda a, b: np.greater(a, b + self.epsilon)
            self.best = -np.Inf
        self.cooldown_counter = 0
        self.wait = 0
        self.lr_epsilon = self.min_lr * 1e-4

    def on_train_begin(self, logs={}):
        self.reset()

    def on_epoch_end(self, epoch, logs={}):
        logs['lr'] = K.get_value(self.model.optimizer.lr)
        current = logs.get(self.monitor)
        if current is None:
            warnings.warn('Learning Rate Plateau Reducing requires %s available!' %
                          self.monitor, RuntimeWarning)
        else:
            if self.in_cooldown():
                self.cooldown_counter -= 1
                self.wait = 0

            if self.monitor_op(current, self.best):
                self.best = current
                self.wait = 0
            elif not self.in_cooldown():
                if self.wait >= self.patience:
                    old_lr = float(K.get_value(self.model.optimizer.lr))
                    if old_lr > self.min_lr + self.lr_epsilon:
                        new_lr = old_lr * self.factor
                        new_lr = max(new_lr, self.min_lr)
                        K.set_value(self.model.optimizer.lr, new_lr)
                        if self.verbose > 0:
                            print('\nEpoch %05d: reducing learning rate to %s.' % (epoch, new_lr))
                        self.cooldown_counter = self.cooldown
                        self.wait = 0
                self.wait += 1

    def in_cooldown(self):
        return self.cooldown_counter > 0

class BatchHistory(Callback):
	'''Callback that records events
	into a `History` object.
	This callback is automatically applied to
	every Keras model. The `History` object
	gets returned by the `fit` method of models.
	'''
	def __init__(self, save_file, save_every_batch):
		self.save_file = save_file
		self.save_every_batch = save_every_batch

	def on_train_begin(self, logs={}):
	    self.batch = []
	    self.history = {}
	    self.epoch_history = {}
	    self.epoch = 0
	    self.batches_per_epoch = 0

	def on_epoch_end(self, epoch, logs={}):
		self.epoch +=1
		if not self.batches_per_epoch:
			self.batches_per_epoch = len(self.batch)

		for k, v in logs.items():
			# use prev value and end with new value
			history_values = [v]* (self.batches_per_epoch)
			
			self.epoch_history.setdefault(k, [0]*self.batches_per_epoch).extend(history_values)

		lr = K.get_value(self.model.optimizer.lr)
		self.epoch_history.setdefault('learning_rate', []).extend([lr] *self.batches_per_epoch)
		if self.epoch > 0:
			self.save()

	def on_batch_end(self, batch, logs={}):
		# increase the epoch on batch 0
		self.batch.append((len(self.batch), self.epoch))
		for k, v in logs.items():
			self.history.setdefault(k, []).append(v)
			lr = K.get_value(self.model.optimizer.lr)
		self.history.setdefault('learning_rate', []).append(lr)
		if (len(self.batch) > 0 and len(self.batch) % self.save_every_batch == 0):
			self.save()

	def save(self):
		data = {
			'batch' : self.batch,
			'history' : self.history,
			'epoch_history': self.epoch_history,
			'params' : self.params
		}
		with open(self.save_file, 'w') as f:
			pickle.dump(data, f)


def create_displacement_accuracy_metric(i, threshold):
	def accuracy_metric(y, pred):
		diff = y[:,i]-pred[:,i]
		abs_diff = K.abs(diff)
		passes = K.lesser(abs_diff, threshold)
		passes = K.cast(passes, 'float32')
		accuracy = K.mean(passes)
		return accuracy

	accuracy_metric.__name__ = 'accuracy_metric_' + str(i)
	return accuracy_metric

def get_regression_learning_params(train_db, params):
	optimizer = params['optimizer']
	loss = params['loss']
	metrics = params['metrics']

	thresholds = np.array([4,4])
	data_min = train_db.get('data_min')
	data_max = train_db.get('data_max')
	low = train_db.get('low')
	high = train_db.get('high')
	target_range = high - low
	data_range = data_max-data_min
	scaling = data_range/target_range
	scaled_thresholds = thresholds/scaling

	for i in range(len(scaled_thresholds)):
		threshold = scaled_thresholds[i]
		metrics.append(create_displacement_accuracy_metric(i, threshold))
	return optimizer, loss, metrics	

def model_2_lane_6_anchor_regression(train_db, params):
	"""
	Creates the model for predicting the displacements from image center of the x coordinates for the 3 anchors for the left lane
	and the 3 anchors for the right lane.  The y location of these coordinates is a hardcorded value for each of the 3 anchors for 
	each lane spline.s
	"""
	optimizer, loss, metrics = get_regression_learning_params(train_db, params)
	input_tensor = Input(shape=(224, 224, 3))
	base_model = ResNet50(weights='imagenet', include_top=False, input_tensor=input_tensor)
	fc_init = params['fc_init']
	x = base_model.output
	x = Flatten()(x)
	x = Dense(6, init=fc_init, name='fc5')(x)

	model = Model(input_tensor, x)
	for layer in model.layers[:len(model.layers)-1]:
	    layer.trainable = False

	model.compile(optimizer=optimizer,
		loss=loss,
		metrics=metrics)

	return model

def model_2_lane_2_anchor_regression(train_db, params):
	"""
	Creates the model for predicting the displacements from image center of the x coordinates that the left and right lane intersect
	the bottom of the image
	"""
	optimizer, loss, metrics = get_regression_learning_params(train_db, params)
	input_tensor = Input(shape=(224, 224, 3))
	base_model = ResNet50(weights='imagenet', include_top=False, input_tensor=input_tensor)
	fc_init = params['fc_init']
	x = base_model.output
	x = Flatten()(x)
	x = Dense(2, init=fc_init, name='fc5')(x)

	model = Model(input_tensor, x)
	for layer in model.layers[:len(model.layers)-1]:
	    layer.trainable = False
	
	model.compile(optimizer=optimizer,
		loss=loss,
		metrics=metrics)	    
	return model	

def model_2_lane_pixel_classification(train_db, params):
	"""
	Represents the bottom of the image as 224 possible classes for the image.  Having a high value
	in label 0 represents a high confidence there is a lane intersecting the bottom of the image at
	pixel 0
	"""
	optimizer = params['optimizer']
	loss = params['loss']
	metrics = params['metrics']

	input_tensor = Input(shape=(224, 224, 3))
	base_model = ResNet50(weights='imagenet', include_top=False, input_tensor=input_tensor)
	fc_init = params['fc_init']
	x = base_model.output
	x = Flatten()(x)
	x = Dense(224, init=fc_init, name='fc5')(x)
	x = Activation('sigmoid')(x)


	model = Model(input_tensor, x)
	for layer in model.layers[:len(model.layers)-1]:
	    layer.trainable = False
	
	model.compile(optimizer=optimizer,
		loss=loss,
		metrics=metrics)	    
	return model

MODEL_FNS = {
	'model_2_lane_6_anchor_regression' : model_2_lane_6_anchor_regression,
	'model_2_lane_2_anchor_regression' : model_2_lane_2_anchor_regression,
	'model_2_lane_pixel_classification' : model_2_lane_pixel_classification
}
def data_generator(db, batch_size=32, num_records=None, shuffle=True):

	batch_xs = []
	batch_ys = []

	if num_records is None:
		num_records = db.get('num_records')
	# make num_records a multiple of batch size
	keys = range(num_records)
	keys = [str(key) for key in keys]
	if shuffle:
		random.shuffle(keys)
	while True:
		#for k, v in it.islice(db.items(),0, 512):
		for v in db.get_keys(keys):
			image_data = v['image_data']
			image_targets = v['image_targets']
			batch_ys.append(image_targets)
			batch_xs.append(image_data)
			if len(batch_xs) >= batch_size:
				yield (np.array(batch_xs), np.array(batch_ys))
				batch_xs = []
				batch_ys = []

def train_model(model_history, train_db, test_db, params):
	nb_epoch = params['nb_epoch']
	batch_size = params['batch_size']
	model_arch = params['model_arch']
	if model_arch not in MODEL_FNS:
		raise Exception ('Model Architecture "' + model_arch + '" does not exist.')
	model_fn =MODEL_FNS[model_arch]
	model = model_fn(train_db, params)


	num_train_records = int(train_db.get('num_records'))
	if 'num_train' in params and params['num_train'] is not None:
		num_train_records = params['num_train']

	num_test_records = int(test_db.get('num_records'))
	if 'num_test' in params and params['num_test'] is not None:
		num_test_records = params['num_test']	

	train_generator = data_generator(train_db, num_records=num_train_records, batch_size=batch_size, shuffle=True)
	test_generator = data_generator(test_db, num_records=num_test_records, batch_size=batch_size)
	
	lr_changer = ReduceLROnPlateau(factor=.5, patience=1, mode='min',min_lr=0.000001)
	history = BatchHistory(model_history.training_history, 100)
	checkpoint_save = model_history.model_save + '.{epoch:02d}-{val_loss:.2f}.hdf5'	
	checkpointer = ModelCheckpoint(filepath=checkpoint_save, verbose=1, save_best_only=True)
	tensorboard = TensorBoard(log_dir=model_history.tensorboard, write_graph=False, histogram_freq=1)
	model.fit_generator(train_generator, num_train_records, nb_epoch, 
		validation_data=test_generator, 
		nb_val_samples=num_test_records,
		callbacks=[history, checkpointer, tensorboard, lr_changer])

	model.save(model_history.model_save + '.hdf5')
	return model

def generate_predictions(model_history, model, train_db, test_db, params):
	batch_size = params['batch_size']
	num_train_records = int(train_db.get('num_records'))
	if 'num_train' in params and params['num_train'] is not None:
		num_train_records = params['num_train']

	num_test_records = int(test_db.get('num_records'))
	if 'num_test' in params and params['num_test'] is not None:
		num_test_records = params['num_test']	

	train_generator = data_generator(train_db, num_records=num_train_records, batch_size=1)
	test_generator = data_generator(test_db, num_records=num_test_records, batch_size=1)


	train_predictions = model.predict_generator(train_generator, num_train_records)
	test_predictions = model.predict_generator(test_generator, num_test_records)

	data = {'train' : train_predictions, 'test' : test_predictions}
	with open(model_history.predictions, 'w') as f:
		pickle.dump(data, f)

def main(args):
	"""
	Every time a model is run, a new line is added to the notes json, recording the model name, time, params, user notes. name of data file used.
	A folder with model name is created.  In that folder will reside checkpoints.  Final Models. repeated notes. predicted train and test results.

	"""
	# load data


	notes = raw_input("Please describe whats going on with this model: ")

	train_db = Simplelmdb(args.data + '/train.lmdb')
	test_db = Simplelmdb(args.test+ '/train.lmdb')	
	model_arch = args.model_arch
	num_train = args.num_train
	num_test = args.num_test	
	meta = {
		'notes' : notes,
		'data' : args.data,
		'num_train' : num_train,
		'num_test' : num_test
	}
	params = {
		'model_arch' : model_arch,
		'num_train' : num_train,
		'num_test' : num_test,
		'fc_init' : args.fc_init,
		'optimizer' : args.optimizer,
		'loss' : args.loss,
		'metrics' : [args.metric],
		'nb_epoch' : args.num_epochs,
		'batch_size' : 32
	}


	model_history  = ModelHistory(args.history, meta, params)
	model = train_model(model_history, train_db, test_db, params)
	generate_predictions(model_history, model, train_db, test_db, params)


if __name__ == '__main__':
	main(args)